### ISSUE TITLE
Fine-tuning 관련 질문 ( XLM-RoBERT-large 추가 )

### ISSUE THREAD

**Issue by tmtmaj** [monologg/KoELECTRA#19]
안녕하세요. 

먼저 공개해주신 코드 유용하게 사용하고 있습니다. 정말 감사합니다. 

XLM-RoBERTa-large 모델의 Fine-tuning 관련해서 질문이 있어 글을 남기게 되었습니다.

공개해 주신 코드는 XLM-RoBERTa-base까지 다루지만 추가적으로 실험을 하고 싶어 config와 run_*.py 코드를 약간 수정하여 XLM-RoBERTa-large 모델에 적용해보았습니다.

다른 task에서는 XLM-RoBERTa-base 이상의 성능을 얻었는데, korquad 에서 터무니 없게 거의 0에 가까운 결과가 나왔습니다.

수정 사항은 다음과 같습니다.

1. config/korquad/xlm-robert-large.json을 추가하였고, 내용은 XLM-RoBERT-base.json에서 모델을 xlm-robert-large로만 바꾸었고 동일한 셋팅을 유지했습니다.

`{
  "task": "korquad",
  "data_dir": "data",
  "ckpt_dir": "ckpt",
  "train_file": "KorQuAD_v1.0_train.json",
  "predict_file": "KorQuAD_v1.0_dev.json",
  "threads": 4,
  "version_2_with_negative": false,
  "null_score_diff_threshold": 0.0,
  "max_seq_length": 512,
  "doc_stride": 128,
  "max_query_length": 64,
  "max_answer_length": 30,
  "n_best_size": 20,
  "verbose_logging": true,
  "overwrite_output_dir": true,
  "evaluate_during_training": true,
  "eval_all_checkpoints": true,
  "save_optimizer": false,
  "do_lower_case": false,
  "do_train": true,
  "do_eval": false,
  "num_train_epochs": 7,
  "weight_decay": 0.0,
  "gradient_accumulation_steps": 1,
  "adam_epsilon": 1e-8,
  "warmup_proportion": 0,
  "max_steps": -1,
  "max_grad_norm": 1.0,
  "no_cuda": false,
  "model_type": "xlm-roberta-large",
  "model_name_or_path": "xlm-roberta-large",
  "output_dir": "xlm-roberta-large-korquad-ckpt",
  "seed": 42,
  "train_batch_size": 4,
  "eval_batch_size": 16,
  "logging_steps": 4000,
  "save_steps": 4000,
  "learning_rate": 5e-5
}
`

2. run_squad.py 에서는 token_type_ids를 제거하는 부분에 xlm-roberta-large를 추가했습니다.

`if args.model_type in ["xlm", "roberta", "distilbert", "distilkobert", "xlm-roberta", "xlm-roberta-large"]:
                del inputs["token_type_ids"]
`
3. /src/util.py 에서 CONFIG_CLASSES, TOKENIZER_CLASSES, MODEL_FOR_QUESTION_ANSWERING
에 각각 "xlm-roberta-large": AutoConfig, "xlm-roberta-large":AutoTokenizer, "xlm-roberta-large": AutoModelForQuestionAnswering를 추가했습니다.

모델의 evaluation 결과입니다.

`12/24/2020 11:49:46 - INFO - __main__ -   ***** Official Eval results *****
12/24/2020 11:49:46 - INFO - __main__ -     official_exact_match = 0.05195704883                                                                                                             962591
12/24/2020 11:49:46 - INFO - __main__ -     official_f1 = 6.183122770846761
12/24/2020 11:49:47 - INFO - __main__ -     HasAns_exact = 0.05195704883962591
12/24/2020 11:49:47 - INFO - __main__ -     HasAns_f1 = 1.1654331969028693
12/24/2020 11:49:47 - INFO - __main__ -     HasAns_total = 5774
12/24/2020 11:49:47 - INFO - __main__ -     best_exact = 0.05195704883962591
12/24/2020 11:49:47 - INFO - __main__ -     best_exact_thresh = 0.0
12/24/2020 11:49:47 - INFO - __main__ -     best_f1 = 1.1654331969028693
12/24/2020 11:49:47 - INFO - __main__ -     best_f1_thresh = 0.0
12/24/2020 11:49:47 - INFO - __main__ -     exact = 0.05195704883962591
12/24/2020 11:49:47 - INFO - __main__ -     f1 = 1.1654331969028693
`

training 시 입력이 제대로 들어오는 지도(tokenizer 문제로 전부 [UNK]로 convert되는 등의 문제) 확인했으나 문제가 없었습니다. 

혹시 어떤 부분에서 문제가 발생할 수 있는지 예상 가시는 부분이 있나요?

감사합니다.

**Comment by monologg** [comment]
원래 large가 finetuning시 수렴이 잘 안되는 이슈가 존재합니다. learning rate와 seed를 조절하면서 우선적으로 테스트해보셔야 할 것 같습니다

**Comment by monologg** [comment]
여담으로 https://huggingface.co/xlm-roberta-large/blob/main/config.json 를 보시면 `"model_type": "xlm-roberta"` 이어서 아래와 같이 하지 않으셔도 될껍니다

```
if args.model_type in ["xlm", "roberta", "distilbert", "distilkobert", "xlm-roberta", "xlm-roberta-large"]: 
    del inputs["token_type_ids"]
```

**Comment by monologg** [comment]
`xlm-roberta-base` 로도 question answering을 돌리셔서 성능이 나오는지도 체크가 필요할 것 같습니다

**Comment by tmtmaj** [comment]
빠른 답변 감사합니다!
`xlm-roberta-base`도 끝까지 훈련해보진 않았는데 먼저 해봐야겠네요. 

`xlm-roberta-base`의 8000 step 결과가 이렇다면 순조롭게 훈련되고 있다고 봐도 괜찮을까요?
` official_exact_match = 17.786629719431936
 official_f1 = 76.30344864490911`

'koelectra-small-v3'의 1000 step 결과 
` official_exact_match = 74.6795981988223
 official_f1 = 85.82023641679791`
이 결과에 비해서는 한참 떨어져보이긴 하네요. 

classification task에서는 정상적으로 수렴하는데 QA task에서 제대로 fine-tuning이 안되는데 신기하기도하고 의문스럽네요. 말씀해주신대로 lr과 seed 조절해보겠습니다. 감사합니다!!

**Comment by monologg** [comment]
> 빠른 답변 감사합니다!
> `xlm-roberta-base`도 끝까지 훈련해보진 않았는데 먼저 해봐야겠네요.
> 
> `xlm-roberta-base`의 8000 step 결과가 이렇다면 순조롭게 훈련되고 있다고 봐도 괜찮을까요?
> ` official_exact_match = 17.786629719431936 official_f1 = 76.30344864490911`
> 
> 'koelectra-small-v3'의 1000 step 결과
> ` official_exact_match = 74.6795981988223 official_f1 = 85.82023641679791`
> 이 결과에 비해서는 한참 떨어져보이긴 하네요.
> 
> classification task에서는 정상적으로 수렴하는데 QA task에서 제대로 fine-tuning이 안되는데 신기하기도하고 의문스럽네요. 말씀해주신대로 lr과 seed 조절해보겠습니다. 감사합니다!!

`xlm-roberta-base` 에서 정상적으로 나온다면 코드 자체의 이슈라기보다는 `xlm-roberta-large`의 수렴 이슈일 가능성이 높겠네요!
lr과 seed 확인 부탁드려요:)
우선은 이슈 닫도록 하겠습니다. 이슈가 추가적으로 생기면 자유롭게 다시 열어주시면 됩니다.
